napari/components/viewer_model.py

```diff
@@ -846,21 +851,22 @@ def _add_layers_with_plugins(
 
         # add each layer to the viewer
         added: List[Layer] = []  # for layers that get added
+        mthd = f'{type(self).__module__}.{type(self).__name__}.open'
+        plugin = hookimpl.plugin_name if hookimpl else None
         for data, filename in zip(layer_data, filenames):
             basename, ext = os.path.splitext(os.path.basename(filename))
             _data = _unify_data_and_user_kwargs(
                 data, kwargs, layer_type, fallback_name=basename
             )
             # actually add the layer
-            new = self._add_layer_from_data(*_data)
-            # some add_* methods return a List[Layer], others just a Layer
-            # we want to always return a list
-            added.extend(new if isinstance(new, list) else [new])
+            for i in self._add_layer_from_data(*_data):
+                i._source = Source(path=filename, plugin=plugin, method=mthd)
+                added.append(i)
         return added
 
     def _add_layer_from_data(
```

Comment:
Is there any way we could put the adding of the source inside this method, or maybe we need an additional method which both adds source and calls this. I think it would be nice if we could have one place in the viewer where the source gets added

Reply:
That’s what I meant in my comment in the original post:

> But a future PR could condense all of this in Layer.__init__ by inspecting the frame stack for the highest level napari API that initiated layer creation.

I can do that here if you want, but I think @jni wanted to keep this simple for the first go round?

Reply:
hmm i guess stack inspection is the most powerful, I was just thinking an additional function or something, but yeah, find to wait on that for this PR and leave this as is now if that might be coming later

Reply:
I think @jni comment about performance is enough to revert back to the earlier commit, and deal with the ugly code for now